hibernate-core/src/main/java/org/hibernate/boot/model/relational/SqlStringGenerationContext.java

```diff
@@ -25,9 +25,25 @@ public interface SqlStringGenerationContext {
 	 * @return The helper for dealing with identifiers in the current JDBC environment.
 	 * <p>
 	 * Note that the Identifiers returned from this helper already account for auto-quoting.
+	 *
+	 * @deprecated Use {@link #toIdentifier(String)} instead.
 	 */
+	@Deprecated
 	IdentifierHelper getIdentifierHelper();
 
+	/**
+	 * Generate an Identifier instance from its simple name as obtained from mapping
+	 * information.
+	 * <p/>
+	 * Note that Identifiers returned from here may be implicitly quoted based on
+	 * 'globally quoted identifiers' or based on reserved words.
+	 *
+	 * @param text The text form of a name as obtained from mapping information.
+	 *
+	 * @return The identifier form of the name.
+	 */
+	Identifier toIdentifier(String text);
+
```

Comment:
How about this instead? Also, is it really necessary to deprecate the `getIdentifierHelper` method?

```suggestion
	 */
	IdentifierHelper getIdentifierHelper();

	/**
	 * Generate an Identifier instance from its simple name as obtained from mapping
	 * information.
	 * <p/>
	 * Note that Identifiers returned from here may be implicitly quoted based on
	 * 'globally quoted identifiers' or based on reserved words.
	 *
	 * @param text The text form of a name as obtained from mapping information.
	 *
	 * @return The identifier form of the name.
	 */
	default Identifier toIdentifier(String text) {
		return getIdentifierHelper() != null ? getIdentifierHelper().toIdentifier( text ) : Identifier.toIdentifier( text );
	}

```

Reply:
> Just for confirmation, the issue is that getIdentifierHelper() returns null which is why it replaces the placeholder with an empty string, right?

No, actually `getIdentifierHelper()` wasn't used at all before this PR. I believe it was introduced by mistake, and recently at that (in my PR that delayed the resolution of default catalog/schema until session factory creation).

The actual fix is the fact that we call `context.catalogWithDefault`/`context.schemaWithDefault` in `injectCatalogAndSchema`. It just turns out that we need to turn a `String` into an `Identifier` in order to do that, so I had to add such a feature to `SqlStringGenerationContext`, and I noticed problems with `getIdentifierHelper`.

> How about this instead? Also, is it really necessary to deprecate the getIdentifierHelper method?

Necessary, no. But as stated above, I believe this method was introduced by mistake. There are no identified use cases for that method, and it can return null in some cases, which is not documented. I'd rather not expose things that we don't